### PR TITLE
TNO-523 Fix AV editing

### PR DESCRIPTION
**Pull request by Fosol**
Resolved creating clips and joining clips when creating a new snippet.

I've simplified the `Clip` and `Join` endpoints, and corrected initial issues with their implementation.

I've also added an 'ugly' way to shrink the top details section of the form.  I was getting annoyed with having to scroll so much to test.

Hide the `Delete` button on new content.

Changed the `Create Snippet` and `Update Snippet` button to `Save`.  If the client doesn't understand we can put it back, but it's annoying to have large buttons that change text for no apparent value.

I've updated the `GridTable` component to handle setting the active row through an optional `assessor` name.

> There are issues with the combined list view when changing to the `Transcript` and `Clips` tab.  @jdtoombs I'll get you resolve this type of issue.  I spent an hour getting no where on the CSS.

> Note there are other bugs that are not part of this PR

image

### CODE REVIEW

File: api/net/Areas/Editor/Controllers/StorageController.cs [hunk 1]
```diff
@@ -191,68 +191,50 @@ public IActionResult Move(string path, string destination, [FromRoute] string? l
...
-    /// <param name="directory"></param>
+    /// <param name="path"></param>
     /// <param name="start"></param>
     /// <param name="end"></param>
-    /// <param name="clipNbr"></param>
-    /// <param name="prefix"></param>
+    /// <param name="outputName"></param>
     /// <returns></returns>
-    [HttpGet("clip")]
+    [HttpPost("clip")]
```
Comment (Fosol) [inline api/net/Areas/Editor/Controllers/StorageController.cs:210]: HTTP GET shouldn't ever be used for introducing any type of 'change'.  We're creating a new file, it should be a POST.

File: api/net/Areas/Editor/Controllers/StorageController.cs [hunk 2]
```diff
@@ -191,68 +191,50 @@ public IActionResult Move(string path, string destination, [FromRoute] string? l
...
     [Produces("application/json")]
-    [ProducesResponseType(typeof(FolderModel), (int)HttpStatusCode.OK)]
+    [ProducesResponseType(typeof(ItemModel), (int)HttpStatusCode.OK)]
```
Comment (Fosol) [inline api/net/Areas/Editor/Controllers/StorageController.cs:212]: Return the file that was created.

File: api/net/Areas/Editor/Controllers/StorageController.cs [hunk 3]
```diff
@@ -191,68 +191,50 @@ public IActionResult Move(string path, string destination, [FromRoute] string? l
...
-    /// <param name="fileName"></param>
-    /// <param name="directory"></param>
+    /// <param name="path"></param>
     /// <param name="prefix"></param>
     /// <returns></returns>
-    [HttpPut("join")]
+    [HttpPost("join")]
     [Produces("application/json")]
-    [ProducesResponseType(typeof(FolderModel), (int)HttpStatusCode.OK)]
+    [ProducesResponseType(typeof(ItemModel), (int)HttpStatusCode.OK)]
```
Comment (Fosol) [inline api/net/Areas/Editor/Controllers/StorageController.cs:230]: Return the file that was created.

File: api/net/Areas/Editor/Controllers/StorageController.cs [hunk 4]
```diff
@@ -191,68 +191,50 @@ public IActionResult Move(string path, string destination, [FromRoute] string? l
...
     /// <summary>
     /// Make a clip from the target file identified in the clip parameter 'prefix'.
     /// </summary>
-    /// <param name="fileName"></param>
-    /// <param name="directory"></param>
+    /// <param name="path"></param>
     /// <param name="prefix"></param>
     /// <returns></returns>
-    [HttpPut("join")]
+    [HttpPost("join")]
```
Comment (Fosol) [inline api/net/Areas/Editor/Controllers/StorageController.cs:228]: HTTP GET shouldn't ever be used for introducing any type of 'change'.  We're creating a new file, it should be a POST.

File: api/net/Areas/Editor/Controllers/StorageController.cs [hunk 5]
```diff
@@ -191,68 +191,50 @@ public IActionResult Move(string path, string destination, [FromRoute] string? l
...
-        var path = _options.CapturePath + directory;
-        var process = FfmpegHelper.GetClipProcess(fileName, path, start, end, clipNbr, prefix);
-
-        process.Start();
-        process.WaitForExit();
-
-        var listing = new FolderModel(_options.CapturePath, directory);
-        return new JsonResult(listing);
+        var safePath = Path.Combine(_options.CapturePath, path.MakeRelativePath());
+        var file = await FfmpegHelper.CreateClipAsync(safePath, start, end, outputName);
```
Comment (Fosol) [inline api/net/Areas/Editor/Controllers/StorageController.cs:218]: Move logic to helper.

File: api/net/Areas/Editor/Controllers/StorageController.cs [hunk 6]
```diff
@@ -191,68 +191,50 @@ public IActionResult Move(string path, string destination, [FromRoute] string? l
...
-
-        process.Start();
-        process.WaitForExit();
-
-        System.IO.File.Delete(muxFile);
-
-        var listing = new FolderModel(_options.CapturePath, directory);
-        return new JsonResult(listing);
+        var safePath = Path.Combine(_options.CapturePath, path.MakeRelativePath());
+        var file = await FfmpegHelper.JoinClipsAsync(safePath, prefix);
```
Comment (Fosol) [inline api/net/Areas/Editor/Controllers/StorageController.cs:236]: Move logic to helper.

File: api/net/Areas/Editor/Models/Storage/FolderModel.cs [hunk 7]
```diff
@@ -40,12 +40,10 @@ public FolderModel(string rootPath, string path)
         var files = System.IO.Directory.GetFileSystemEntries(safePath);
         Array.Sort(files, string.CompareOrdinal);
 
-        foreach (var fullName in files)
+        // Don't include temporary folder.
+        foreach (var fullName in files.Where(n => new DirectoryInfo(n).Name != "_tmp"))
         {
-            if (fullName.FileExists() || fullName.DirectoryExists())
```
Comment (Fosol) [inline api/net/Areas/Editor/Models/Storage/FolderModel.cs:45]: I don't know why we need to validate the file/folder exists?  Line 40 just fetched them, which means they exist.

File: api/net/Areas/Editor/Models/Storage/ItemModel.cs [hunk 8]
```diff
@@ -64,6 +64,11 @@ public ItemModel(string path)
             this.MimeType = MimeTypeMap.GetMimeType(this.Extension);
             this.Modified = info.LastWriteTime;
         }
+        else
+        {
+            var info = new System.IO.DirectoryInfo(path);
+            this.Modified = info.LastWriteTime;
```
Comment (Fosol) [inline api/net/Areas/Editor/Models/Storage/ItemModel.cs:70]: Get the modified date for folders.

File: app/editor/src/features/content/form/ClipDirectoryTable.tsx [hunk 9]
```diff
@@ -28,15 +28,21 @@ export const ClipDirectoryTable: React.FC<IClipDirectoryTableProps> = ({
...
       <GridTable
+        activeId={activeId}
         paging={{ showPaging: false }}
         columns={clipDirectoryColumns(onDelete, onSelect, onDownload, onAttach, values)}
         header={ClipDirectoryFilter}
         data={data}
         className="file-table"
-        onRowClick={(row) => navigate(row.original)}
+        onRowClick={(row) => {
+          setActiveId(row.index);
```
Comment (Fosol) [inline app/editor/src/features/content/form/ClipDirectoryTable.tsx:43]: This doesn't work yet.  I need to modify the GridTable to allow for either number or string to identify a row.  Another day another PR.

File: app/editor/src/features/content/form/ContentClipForm.tsx [hunk 10]
```diff
@@ -192,6 +189,17 @@ export const ContentClipForm: React.FC<IContentClipFormProps> = ({
               Set End
             </Button>
           </Col>
+          <Col>
```
Comment (Fosol) [inline app/editor/src/features/content/form/ContentClipForm.tsx:192]: Moved buttons around based on workflow efficiencies.

File: app/editor/src/features/content/form/ContentForm.tsx [hunk 11]
```diff
@@ -190,118 +191,159 @@ export const ContentForm: React.FC<IContentFormProps> = ({ contentType = Content
...
-                          props.setFieldValue('otherSource', value);
-                          props.setFieldValue('source', value);
-                          if (!!value) {
-                            props.setFieldValue('dataSourceId', undefined);
-                          }
-                        }}
-                        required={!!props.values.otherSource}
-                      />
+                <Row alignItems="flex-start" className="content-details">
+                  <Show visible={size === 0}>
```
Comment (Fosol) [inline app/editor/src/features/content/form/ContentForm.tsx:195]: Just a simple way to get more vertical space back on the screen when editing clips.

File: app/editor/src/features/content/form/ContentTranscriptForm.tsx [hunk 12]
```diff
@@ -17,7 +17,6 @@ export const ContentTranscriptForm: React.FC = () => {
         label="Transcript"
         value={values.transcription}
         onChange={(e: any) => setFieldValue('transcription', e.target.value)}
-        style={{ width: '1000px', height: '500px' }}
```
Comment (Fosol) [inline app/editor/src/features/content/form/ContentTranscriptForm.tsx:20]: Causing display issues on the combined view.

File: app/editor/src/features/content/form/constants/clipDirectoryColumns.tsx [hunk 13]
```diff
@@ -44,44 +45,56 @@ export const clipDirectoryColumns = (
...
-        />
-        <FaTrash
-          className="delete fa-lg"
-          title="Delete"
-          onClick={() => {
-            onDelete(row.values);
-          }}
-        />
-      </div>
+      <Row className={`file-actions ${row.values.isDirectory ? 'hidden' : 'center'}`} wrap="nowrap">
```
Comment (Fosol) [inline app/editor/src/features/content/form/constants/clipDirectoryColumns.tsx:66]: Simplified the CSS.  I'd like to be able to remove the `fa-lg` classname values too at some point.

File: app/editor/src/features/content/list-view/constants/columns.tsx [hunk 14]
```diff
@@ -1,65 +1,73 @@
...
+      maxWidth: 20,
+      accessor: (row) =>
+        row.status === ContentStatusName.Publish || row.status === ContentStatusName.Published,
+      Cell: ({ value }: { value: boolean }) => {
+        return <Checkbox checked={value} />;
+      },
+    },
+  ];
+
+  if (!minimize) {
```
Comment (Fosol) [inline app/editor/src/features/content/list-view/constants/columns.tsx:56]: Only include if full screen mode.

File: libs/net/dal/Helpers/FfmpegHelper.cs [hunk 15]
```diff
@@ -1,124 +1,152 @@
...
-    /// <param name="end"></param>
-    /// <param name="clipNbr"></param>
-    /// <param name="prefix"></param>
+    /// <param name="path">Full path to the source file.</param>
+    /// <param name="start">Starting position in file.</param>
+    /// <param name="end">Ending position in file.</param>
+    /// <param name="outputName">The name of the file to create.</param>
     /// <returns>System.Diagnostics.Process</returns>
-    public static System.Diagnostics.Process GetClipProcess(string fileName, string directory, int start, int end, int clipNbr, string prefix)
+    public static async Task<string> CreateClipAsync(string path, int start, int end, string outputName)
```
Comment (Fosol) [inline libs/net/dal/Helpers/FfmpegHelper.cs:20]: Simplified parameters to only require what is needed to clip.

File: libs/net/dal/Helpers/FfmpegHelper.cs [hunk 16]
```diff
@@ -1,124 +1,152 @@
...
+    }
 
-        return process;
+    /// <summary>
+    /// Run the ffmpeg command and create a new clip by joining all clips matching the prefix and source file extension type.
+    /// </summary>
+    /// <param name="path">Full path to the source file.</param>
+    /// <param name="prefix">Filename prefix to search for files to join.</param>
+    /// <returns></returns>
+    public static async Task<string> JoinClipsAsync(string path, string prefix)
```
Comment (Fosol) [inline libs/net/dal/Helpers/FfmpegHelper.cs:36]: Simplified parameters to only require what is needed to join based on convention.  At some point I would like to create a join that takes an array of files, but for now I kept the original functionality.

File: libs/net/dal/Helpers/FfmpegHelper.cs [hunk 17]
```diff
@@ -1,124 +1,152 @@
...
-    /// <param name="end"></param>
-    /// <param name="clipNbr"></param>
-    /// <param name="prefix"></param>
+    /// <param name="path">Full path to the source file.</param>
+    /// <param name="start">Starting position in file.</param>
+    /// <param name="end">Ending position in file.</param>
+    /// <param name="outputName">The name of the file to create.</param>
     /// <returns></returns>
-    public static string GenerateClipCommand(string fileName, string directory, int start, int end, int clipNbr, string prefix)
+    private static (string Cmd, string Output) GenerateClipCommand(string path, int start, int end, string outputName)
```
Comment (Fosol) [inline libs/net/dal/Helpers/FfmpegHelper.cs:57]: With a tuple we can return both the cmd and the output file.

File: libs/net/dal/Helpers/FfmpegHelper.cs [hunk 18]
```diff
@@ -1,124 +1,152 @@
...
-    /// <param name="fileName"></param>
-    /// <param name="path"></param>
+    /// <param name="path">Full path to the source file.</param>
     /// <param name="muxFile"></param>
-    /// <param name="format"></param>
-    /// <param name="prefix"></param>
+    /// <param name="prefix">Filename prefix to search for files to join.</param>
     /// <returns></returns>
-    public static System.Diagnostics.Process GetJoinProcess(string directory, string fileName, string path, string muxFile, string format, string prefix)
+    private static (string Cmd, string Output) GenerateJoinCommand(string path, string muxFile, string prefix)
```
Comment (Fosol) [inline libs/net/dal/Helpers/FfmpegHelper.cs:77]: With a tuple we can return both the cmd and the output file.

File: libs/net/dal/Helpers/FfmpegHelper.cs [hunk 19]
```diff
@@ -1,124 +1,152 @@
...
+    /// <param name="path">Full path to the source file.</param>
+    /// <param name="prefix">Filename prefix to search for files to join.</param>
+    /// <returns></returns>
+    private static string GenerateMuxfile(string path, string prefix)
+    {
+        var ext = Path.GetExtension(path).Replace(".", "");
+        var directory = Path.GetDirectoryName(path) ?? "";
+        var clips = Directory.GetFileSystemEntries(directory, prefix + "_*." + ext);
+        // TODO: Two users with same prefix will overwrite each other.
+        var muxfile = Path.Combine(directory, "_tmp", $"{prefix}.txt");
```
Comment (Fosol) [inline libs/net/dal/Helpers/FfmpegHelper.cs:98]: The file is placed in the current directory to limit the possibility of overwriting another user's file for now.  Need to handle this better in the future.

File: libs/npm/core/src/components/tabs/TabContainer.tsx [hunk 20]
```diff
@@ -12,7 +12,7 @@ export const TabContainer: React.FC<React.HTMLAttributes<HTMLDivElement>> = ({
   className,
 }) => {
   return (
-    <styled.TabContainer className={`${className ?? 'tab-container'}`}>
+    <styled.TabContainer className={`tab-container${!!className ? ` ${className}` : ''}`}>
```
Comment (Fosol) [inline libs/npm/core/src/components/tabs/TabContainer.tsx:15]: Should always have the default classname.

File: libs/npm/core/src/components/tabs/Tabs.tsx [hunk 21]
```diff
@@ -17,7 +17,7 @@ export interface ITabsProps extends React.HTMLAttributes<HTMLDivElement> {
  */
 export const Tabs: React.FC<ITabsProps> = ({ tabs, children, className }) => {
   return (
-    <styled.Tabs className={`${className ?? 'tabs'}`}>
+    <styled.Tabs className={`tabs${!!className ? ` ${className}` : ''}`}>
```
Comment (Fosol) [inline libs/npm/core/src/components/tabs/Tabs.tsx:20]: Should always have the default classname.

File: app/editor/package.json [hunk 22]
```diff
@@ -1,129 +1,129 @@
...
+    "react-router-dom": "6.3.0",
+    "react-scripts": "5.0.1",
+    "react-select": "5.3.2",
+    "react-table": "7.8.0",
+    "react-text-mask": "5.4.3",
+    "react-toastify": "9.0.2",
+    "react-tooltip": "4.2.21",
+    "redux-logger": "3.0.6",
+    "styled-components": "5.3.5",
+    "tno-core": "0.0.165",
```
Comment (Fosol) [inline app/editor/package.json:42]: Updated tno-core version

File: app/editor/src/features/content/form/ClipDirectoryTable.tsx [hunk 23]
```diff
@@ -28,15 +28,22 @@ export const ClipDirectoryTable: React.FC<IClipDirectoryTableProps> = ({
 }) => {
   const { values } = useFormikContext<IItemModel>();
 
+  const [activeId, setActiveId] = React.useState<string>();
+
   return (
     <styled.ClipDirectoryTable>
       <GridTable
+        activeId={activeId}
+        activeAssessor="name"
```
Comment (Fosol) [inline app/editor/src/features/content/form/ClipDirectoryTable.tsx:37]: We can now highlight the active row based on any assessor value.

File: app/editor/src/features/content/form/ContentClipForm.tsx [hunk 24]
```diff
@@ -86,21 +85,24 @@ export const ContentClipForm: React.FC<IContentClipFormProps> = ({
...
   const joinClips = async () => {
-    await storage.join(currFile, folder.path, prefix).then((data) => {
-      setFolder({ ...data, items: data.items });
-      setClipNbr(1);
+    await storage.join(`${folder.path}/${currFile}`, prefix).then((item) => {
+      setItem(item);
+      setFolder({ ...folder, items: [...folder.items, item] });
       setStart('');
       setEnd('');
+      setCurrFile(!!item ? item.name : '');
```
Comment (Fosol) [inline app/editor/src/features/content/form/ContentClipForm.tsx:102]: Make the active file the new joined file.

File: libs/npm/core/package.json [hunk 25]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "tno-core",
   "description": "TNO shared library",
-  "version": "0.0.164",
+  "version": "0.0.165",
```
Comment (Fosol) [inline libs/npm/core/package.json:4]: Update `GridTable` component

**Review by jdtoombs — APPROVED** [review]

**Pull request merged**